### PR TITLE
refactor: decouple memory management from stream, make it accessible for both batch and streaming

### PR DESCRIPTION
**Pull request by BowenXiao1999**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Main idea:
1. Rename `LruManager` to `GlobalMemoryManager`, move it from `stream` crate to `compute` crate. Can not move to `common` as it depends on `risingwave_stream` and `risingwave_batch`. This comes from what we have discussed in the memory management rfc: https://github.com/risingwavelabs/rfcs/pull/26.

2. Fully decouple `risingwave_stream` and memory manager. Before this pr, streaming executor access to lru manager to create cache. However, this will cause cyclic reference if we move `LruManager` out from `risingwave_stream`. What executor really need is the watermark epoch, so instead of let `risingwave_stream` access to Memory Manager, just store the watermark epoch in the `LocalStreamManager` and when executors are building, they can read this value and then they can create cache with their own. Personally I think this is more clean: memory manager have access to stream/batch two components, and vic versa no.

3. Currently the memory manager ref is not stored anywhere. Thinking of where to store it. 🤔

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/batch/src/task/task_manager.rs [hunk 1]
```diff
@@ -202,6 +202,18 @@ impl BatchManager {
     pub fn config(&self) -> &BatchConfig {
         &self.config
     }
+
+    /// Kill batch queries with larges memory consumption per task. Required to maintain task level
+    /// memory usage in the struct. Will be called by global memory manager.
+    pub fn kill_queries(&self) {
+        todo!()
+    }
```
Comment (BowenXiao1999) [inline src/batch/src/task/task_manager.rs:210]: Just for better understanding, we can add this in future PRs.

File: src/compute/src/server.rs [hunk 2]
```diff
@@ -212,8 +213,32 @@ pub async fn compute_node_serve(
...
+    let batch_mgr_clone = batch_mgr.clone();
+    let stream_mgr_clone = stream_mgr.clone();
+    let create_lru_manager = move || {
+        let mgr = GlobalMemoryManager::new(
+            opts.total_memory_bytes,
+            config.streaming.barrier_interval_ms,
+            streaming_metrics.clone(),
+        );
+        // Run a background memory monitor
+        tokio::spawn(mgr.clone().run(batch_mgr_clone, stream_mgr_clone));
```
Comment (liurenjie1024) [inline src/compute/src/server.rs:228]: It should be a long running task, which trigger the `run` method periodically, but we can refine it later.

File: src/stream/src/executor/hash_agg.rs [hunk 3]
```diff
@@ -158,7 +159,7 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
         group_key_indices: Vec<usize>,
         group_by_cache_size: usize,
         extreme_cache_size: usize,
-        lru_manager: Option<LruManagerRef>,
+        lru_manager: AtomicU64RefOpt,
```
Comment (liurenjie1024) [inline src/stream/src/executor/hash_agg.rs:162]: Rename parameter?

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM

File: src/stream/src/executor/top_n/group_top_n.rs [hunk 4]
```diff
@@ -50,7 +51,7 @@ impl<K: HashKey, S: StateStore, const WITH_TIES: bool> GroupTopNExecutor<K, S, W
         executor_id: u64,
         group_by: Vec<usize>,
         state_table: StateTable<S>,
-        lru_manager: Option<LruManagerRef>,
+        watermark_epoch: AtomicU64RefOpt,
```
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply src/stream/src/executor/top_n/group_top_n.rs:54]: That's good news!

**Review by liurenjie1024 — APPROVED** [review]

**Comment by yuhao-su** [comment]
Can we provide a `StreamMemoryManager` and `BatchMemoryManager` to hide the details of handling memory eviction/kill queries to `GlobalMemoryManager`.

The current impl passes a watermark epoch to the stream and `GlobalMemoryManager`  hold a batch manager. Seem the coupling is too strong.

File: src/compute/src/server.rs (hunk 2 above)
Comment (BowenXiao1999) [reply src/compute/src/server.rs:228]: There is a tick interval in `run` so I guess this is not a big problem?

**Comment by liurenjie1024** [comment]
> > Can we provide a `StreamMemoryManager` and `BatchMemoryManager` to hide the details of handling memory eviction/kill queries to `GlobalMemoryManager`.
> > The current impl passes a watermark epoch to the stream and `GlobalMemoryManager` hold a batch manager. Seem the coupling is too strong.
> 
> I though this is expected? Now indeed the `GlobalMemoryManager` is designed to couple with both Stream and Batch, and yes we can provide two memory manager, But I'm not sure whether there will be some inconsistency problem and can we ensure Batch do not occupy too much memory by Stream if they run on 2 different corountine? This may relates to the algorithm in control, I forgot some of the details in the meeting, so cc @fuyufjh @liurenjie1024 .

Yes, I think it's expected and don't see a strong reason to introduce another layer.

**Pull request merged**